### PR TITLE
[Internal] Distributed tracing: Adds a sample to collect activities and events using custom listener

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description
Sample to collect activities using AppInsight SDK covered [here](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/master/Microsoft.Azure.Cosmos.Samples/Usage/ApplicationInsights)
Sample to collect activities using Azure Montor SDK covered [here](https://github.com/Azure/azure-cosmos-dotnet-v3/tree/master/Microsoft.Azure.Cosmos.Samples/Usage/OpenTelemetry). (All the Open Telemetry Exporter follows the same pattern.)

As part of this PR, adding a sample where customer can write their own listeners to capture activities and events generated by SDK.

#### Why would anybody do that?
1. To write its own monitoring library with their custom implementation.
2. To support an APM tool which is not open telemetry compliant.

_Note: It is a simple sample. Anybody can get as creative as they want to make it better in terms of usability._
## Type of change

Please delete options that are not relevant.

- [] Non Functional

### CODE REVIEW

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        /// </summary>
+        public void OnCompleted() {
+            throw new NotImplementedException(); 
+        }
+
+        /// <summary>
+        /// IObserver Override
+        /// </summary>
+        public void OnError(Exception error) {
+            throw new NotImplementedException();
```
Comment (jcocchi) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs:38]: Can we throw the actual exception instead of a NotImplementedException?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            this.eventSourceName = eventSourceName;
+
+            DiagnosticListener.AllListeners.Subscribe(this);
+        }
+
+        /// <summary>
+        /// IObserver Override
+        /// </summary>
+        public void OnCompleted() {
+            throw new NotImplementedException();
```
Comment (jcocchi) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs:31]: What's an example of something a customer would do in this scenario. Maybe a log statement here, something else?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Content Include="..\..\..\..\..\Users\sourabhjain\.nuget\packages\microsoft.azure.cosmos\3.33.0-preview\contentFiles\any\netstandard2.0\ThirdPartyNotice.txt" Link="ThirdPartyNotice.txt">
```
Comment (jcocchi) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:11]: Is this intended to point to a path on your local machine?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      <Private>False</Private>
+      <NuGetPackageVersion>3.33.0-preview</NuGetPackageVersion>
```
Comment (jcocchi) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:13]: Should this package version match the version on line 21?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj (hunk 3 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:11]: no fixed.

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj (hunk 4 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:13]: no this project uses latest version of cosmosdb

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs (hunk 2 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs:31]: I don't know when it comes here (i didn't try to go deep there also) but added logs to avoid any confusion

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs (hunk 1 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs:38]: Logging the exception

**Review by jcocchi — DISMISSED** [review]
LGTM

**Review by jcocchi — APPROVED** [review]

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/AppSettings.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/AppSettings.json:1]: Check the other Sample CSPROJ files. Don't duplicate this file, but rather reference the existing one

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net6.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.35.2-preview" />
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:11]: Do we need the preview SDK for this?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/Program.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                throw new ArgumentNullException("Please specify a valid CosmosDBEndPointUrl in the appSettings.json");
+            }
+
+            string authKey = configuration["CosmosDBAuthorizationKey"];
+            if (string.IsNullOrEmpty(authKey) || string.Equals(authKey, "Super secret key"))
+            {
+                throw new ArgumentException("Please specify a valid CosmosDBAuthorizationKey in the appSettings.json");
+            }
+
+            using CustomDiagnosticAndEventListener listener
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/Program.cs:32]: Can we add some description of what is the use case for this Sample project? What is it meant for and when would a user reference it?

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj (hunk 6 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.csproj:11]: yes..this feature is in preview right now.

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/Program.cs (hunk 7 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/Program.cs:32]: Already added in listener file

Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/Program.cs:32]: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/833e5bd70c7079338f5ecf21ce590a58ff93a9b4/Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/CustomDiagnosticAndEventListener.cs#L7

File: Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/AppSettings.json (hunk 5 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos.Samples/Usage/CustomDiagnosticAndEventListener/AppSettings.json]: fixed for other related samples also

File: Microsoft.Azure.Cosmos.Samples/Usage/appSettings.json [hunk 8]
```diff
@@ -1,4 +1,12 @@
 {
   "EndPointUrl": "https://localhost:8081",
-  "AuthorizationKey": "Super secret key"
+  "AuthorizationKey": "Super secret key",
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos.Samples/Usage/appSettings.json:3]: Please leave the base file as is. OpenTelemetry and ApplicationInsights projects, if they require special settings, makes sense to have a special version for them. But CustomDiagnosticsAndEventListener does not.

**Review by ealsur — APPROVED** [review]

**Comment by kirankumarkolli** [comment]
Thank you @sourabh1007 for your contributions.
 Can you please help expand on below aspects/clarifications?

- Listeners lifetime: Do they needs to be pre-created before CosmosClient or doesn't matter? 
- Sync vs async invocations: Do the listener block the main line flow in SDK or runs asynchronously?
    - User-case: In which scenarios developers needs to write customer listener
- Subscriptions: Is OnComplete() expected to un-subscribe? 
    - In what scenarios multiple subscriptions are expected? 
- Activities member: Its instance level lifetime (concurrentBag), entries are added with no-eviction right?

**Comment by sourabh1007** [comment]
To answer your question:

**Listeners lifetime: Do they needs to be pre-created before `CosmosClient` or doesn't matter?**
Listener can be initialized anywhere; it will start listening call from that point onwards. Ideal way to initialize it in the starting itself and filter activities which you don't want to record.

**Sync vs async invocations: Do the listener block the main line flow in SDK or runs asynchronously?**
it is all sync calls. So it is important to have performant listener and recommended to use sampling to avoid the impact on p99.

**User-case: In which scenarios developers needs to write customer listener**
Already mention in the PR description:
a) To write its own monitoring library with their custom implementation.
b) To support an APM tool which is not open telemetry compliant. (basically write a custom listener to feed this APM tool)

**Subscriptions: Is OnComplete() expected to un-subscribe?**
ref. https://learn.microsoft.com/en-us/dotnet/api/system.iobserver-1.oncompleted?view=net-7.0

**In what scenarios multiple subscriptions are expected?**
Ideally there should not be multiple listeners in prod because it impacts the overall latency. but there can be a scenario where you want to feed data to AppInsight tool and along with that you want to print on console for debugging. you can use AzureMonitorExporter and ConsoleExporter.
Basically when you want to feed telemetry to multiple places then there will be multiple listeners.

**Activities member: Its instance level lifetime (concurrentBag), entries are added with no-eviction right?**
It is just a sample code. In dispose I am cleaning up this whole list by disposing all the activities.
